src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+        font_shaper_cache: font.ShaperCache,
+
+        /// The images that we may render.
+        images: ImageMap = .{},
+        image_placements: ImagePlacementList = .{},
+        image_bg_end: u32 = 0,
+        image_text_end: u32 = 0,
+        image_virtual: bool = false,
+
+        /// Graphics API state.
+        api: GraphicsAPI,
+
+        /// The CVDisplayLink used to drive the rendering loop in
+        /// sync with the display. This is void on platforms that
+        /// don't support a display link.
+        display_link: ?DisplayLink = null,
+
+        /// Health of the most recently completed frame.
+        health: std.atomic.Value(Health) = .{ .raw = .healthy },
+
+        /// Our swap chain (multiple buffering)
+        swap_chain: SwapChain,
+
+        /// This value is used to force-update swap chain targets in the
+        /// event of a config change that requires it (such as blending mode).
+        target_config_modified: usize = 0,
+
+        /// If something happened that requires us to reinitialize our shaders,
+        /// this is set to true so that we can do that whenever possible.
+        reinitialize_shaders: bool = false,
+
+        /// Whether or not we have custom shaders.
+        has_custom_shaders: bool = false,
+
+        /// Our shader pipelines.
+        shaders: Shaders,
+
+        /// Swap chain which maintains multiple copies of the state needed to
+        /// render a frame, so that we can start building the next frame while
+        /// the previous frame is still being processed on the GPU.
+        const SwapChain = struct {
+            // The count of buffers we use for double/triple buffering.
+            // If this is one then we don't do any double+ buffering at all.
+            // This is comptime because there isn't a good reason to change
+            // this at runtime and there is a lot of complexity to support it.
+            const buf_count = count: {
+                if (@hasDecl(GraphicsAPI, "swap_chain_count")) {
+                    break :count GraphicsAPI.swap_chain_count;
+                }
+
+                // Default to triple buffering if
+                // graphics API has no preference.
+                break :count 3;
+            };
+
+            /// `buf_count` structs that can hold the
+            /// data needed by the GPU to draw a frame.
+            frames: [buf_count]FrameState,
+            /// Index of the most recently used frame state struct.
+            frame_index: std.math.IntFittingRange(0, buf_count) = 0,
+            /// Semaphore that we wait on to make sure we have an available
+            /// frame state struct so we can start working on a new frame.
+            frame_sema: std.Thread.Semaphore = .{ .permits = buf_count },
+
+            /// Set to true when deinited, if you try to deinit a defunct
+            /// swap chain it will just be ignored, to prevent double-free.
+            defunct: bool = false,
```

Comment:
Is this just for runtime safety? I tried spying how this could ever happen under normal scenarios and don't see it.

It's a very lightweight thing for a lot of safety so I'm happy to keep it, just want clarification.

Reply:
Added clarification to this doc comment about that.